### ISSUE TITLE
Throttled writing to flash

### ISSUE THREAD

**Issue by mrquincle** [crownstone/bluenet#86]
Currently, there is a particular function that delays writing to flash, [setDelayed](https://github.com/crownstone/bluenet/blob/master/source/src/storage/cs_State.cpp#L403). This avoids writing to flash multiple times in a row (which would lead to wear, see [Tesla](https://insideevs.com/news/376037/tesla-mcu-emmc-memory-issue/)). For example, when dimming we do not want to store every intermediate state. Only after the user decided upon a value can that value be committed to flash. Every time the value changes we postpone sending it to flash. In theory, if the user keeps changing it all the time, it will never be written to flash! The implementation is done by counters per delayed variable. Every tick each of these counters is decreased. Note: the alternative is a single "semi-clock" (starting from 0) that is incrementing. Then we set values related to the current clock value. The advantage is that we only need to compare per value (rather than write). The disadvantage is that we need to be careful when the clock "wrapping around". The current value is an uint32_t (with milliseconds as resolution) which means wrapping around every 50 days. Often enough to care... If this would be at a resolution of seconds, this would be 136 years. Then, we might use a semi-clock instead.

If we get input from the outside world, we might also want to **throttle**. Rather than only writing to flash **after** a certain time, we want to write **immediately** to flash if this didn't happen in the past period. 

# How to implement this?

+ Per flash variable that requires throttling have a ram variable that stores when this variable has been written to flash last. We can reuse the current queue for this. 
+ In the `setThrottled` function we not only set it in RAM, but we check if the current value in RAM is different before we set it. The result we communicate with an additional parameter to `addToQueue`.
+ We adjust `addToQueue` that if the value is not present (`!found`), we immediately write it to FLASH (only in this mode). We add the field to the queue like usual in this function. 
+ If the value is present, we do not write to FLASH, but we update the `_store_queue[i].execute` boolean field to `true`.
+ Now, in `delayedStoreTick` when the counter becomes zero, the `execute` field is checked. In the delayed mode this is set to true and it will always be written when it becomes zero. In the throttled mode, it is normally set to false. It will not be written when it becomes zero (it just only had as function to prevent other - potentially identical - flash writes). In the throttled mode, if it is true, the value in the mean-time has changed, then it also will be written when the counter reaches zero. 
+ After writing this value in throttled mode, we should not remove it from the queue! The counter has to be set to the original counter value again. This means that we need in the `cs_state_store_queue_t` field besides a bool `execute`, also an `init_counter` value.

# Additional considerations

+ If `setThrottled` is called multiple times with [different] throttle periods, we can update the throttle period (both the `counter` and the `init_counter` variables). If we do this often enough the counter will never reach zero. In other words, apart from the first time, it will never be written to flash. It makes more sense to update the `init_counter` variable here, but not the `counter` variable itself. The updated variable will then be written after `counter` reaches zero, subsequently a new throttle period is used using `init_counter` after writing this value to flash.

Implementation will be on the [store_multiple_values_per_type](https://github.com/crownstone/bluenet/blob/store_multiple_values_per_type/source/src/storage/cs_State.cpp) branch.
